packages/medusa/src/services/totals-new.ts

```diff
@@ -0,0 +1,692 @@
+      return acc + Math.round(lineSubtotal * (next.rate / 100))
+    }, 0)
+
+    return lineSubtotal + taxTotal
+  }
+
+  /**
+   * Calcul and return the gift cards totals
+   * @param giftCardableAmount
+   * @param giftCardTransactions
+   * @param region
+   * @param giftCards
+   */
+  async getGiftCardTotals(
+    giftCardableAmount: number,
+    {
+      giftCardTransactions,
+      region,
+      giftCards,
+    }: {
+      region: Region
+      giftCardTransactions?: {
+        tax_rate: number
+        is_taxable: boolean
+        amount: number
+      }[]
+      giftCards?: GiftCard[]
+    }
+  ): Promise<{
+    total: number
+    tax_total: number
+  }> {
+    if (giftCardTransactions) {
+      return this.getGiftCardTransactionsTotals({
+        giftCardTransactions,
+        region,
+      })
+    }
+
+    const result = {
+      total: 0,
+      tax_total: 0,
+    }
+
+    if (!giftCards?.length) {
+      return result
+    }
+
+    const giftAmount = giftCards.reduce((acc, next) => acc + next.balance, 0)
+    result.total = Math.min(giftCardableAmount, giftAmount)
+
+    if (region?.gift_cards_taxable) {
+      result.tax_total = Math.round((result.total * region.tax_rate) / 100)
+      return result
+    }
+
+    return result
+  }
+
+  /**
+   * Calcul and return the gift cards totals based on their transactions
+   * @param gift_card_transactions
+   * @param region
+   */
+  getGiftCardTransactionsTotals({
+    giftCardTransactions,
+    region,
+  }: {
+    giftCardTransactions: {
+      tax_rate: number
+      is_taxable: boolean
+      amount: number
+    }[]
+    region: { gift_cards_taxable: boolean; tax_rate: number }
+  }): { total: number; tax_total: number } {
+    return giftCardTransactions.reduce(
+      (acc, next) => {
+        let taxMultiplier = (next.tax_rate || 0) / 100
+
+        // Previously we did not record whether a gift card was taxable or not.
+        // All gift cards where is_taxable === null are from the old system,
+        // where we defaulted to taxable gift cards.
+        //
+        // This is a backwards compatability fix for orders that were created
+        // before we added the gift card tax rate.
+        if (next.is_taxable === null && region?.gift_cards_taxable) {
+          taxMultiplier = region.tax_rate / 100
+        }
+
+        return {
+          total: acc.total + next.amount,
+          tax_total: acc.tax_total + next.amount * taxMultiplier,
+        }
+      },
+      {
+        total: 0,
+        tax_total: 0,
+      }
+    )
+  }
+
+  /**
+   * Calcul and return the shipping methods totals for either the legacy calculation or the new calculation
+   * @param shippingMethods
+   * @param includeTax
+   * @param discounts
+   * @param taxRate
+   * @param calculationContext
+   * @param useExistingTaxLines Force to use the tax lines of the shipping method instead of fetching them
+   */
+  async getShippingMethodsTotals(
```

Comment:
thought: Same thought as above with `getLineItemTotals` - can we overload instead?